### PR TITLE
mgr/dashboard: enable different URL for users of browser to Grafana

### PR DESCRIPTION
**Pull request by p-se**
Fixes: https://tracker.ceph.com/issues/47386

Signed-off-by: Patrick Seidensal <pseidensal@suse.com>

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: doc/cephadm/monitoring.rst
Comment (LenzGr) [inline doc/cephadm/monitoring.rst:80]: Hmm, how about this?
```suggestion
Cephadm takes care of the configuration of Prometheus, Grafana, and Alertmanager
```

File: doc/cephadm/monitoring.rst
Comment (LenzGr) [inline doc/cephadm/monitoring.rst:84]:
```suggestion
Dashboard user's browser might not be able to access the URL to Grafana configured in Ceph
```

File: doc/cephadm/monitoring.rst
Comment (LenzGr) [inline doc/cephadm/monitoring.rst:83]:
```suggestion
However, there is one exception to this rule. In some setups, the
```

File: doc/cephadm/monitoring.rst
Comment (LenzGr) [inline doc/cephadm/monitoring.rst:89]:
```suggestion
can be used to configure which URL for accessing Grafana by the user´s browser
```

**Review by votdev — CHANGES_REQUESTED** [review]
Looks good, except for the somewhat bumpy sentences. Please evaluate if the suggested sentences are better.

File: doc/mgr/dashboard.rst
Comment (LenzGr) [inline doc/mgr/dashboard.rst:614]:
```suggestion
working connections are required in order to be able to see Grafana graphs in Ceph
```

File: doc/mgr/dashboard.rst
Comment (LenzGr) [inline doc/mgr/dashboard.rst:617]:
```suggestion
- The backend (Ceph Mgr module) needs to verify the existence of the requested graph. If this
```

File: doc/mgr/dashboard.rst
Comment (LenzGr) [inline doc/mgr/dashboard.rst:619]:
```suggestion
- The frontend then requests the Grafana graphs directly from the user´s
```

File: doc/mgr/dashboard.rst
Comment (LenzGr) [inline doc/mgr/dashboard.rst:620]:
```suggestion
  browser using an iframe. The Grafana instance is accessed directly without any
```

File: doc/mgr/dashboard.rst
Comment (LenzGr) [inline doc/mgr/dashboard.rst:624]:
```suggestion
Now, it might be the case that your environment makes it difficult for the user´s browser
to directly access the Grafana URL configured in Ceph Dashboard. To solve this issue, a
```

File: doc/mgr/dashboard.rst
Comment (LenzGr) [inline doc/mgr/dashboard.rst:626]:
```suggestion
(the user´s browser) which URL it should use to access Grafana. This setting
```

Comment (smithfarm) [reply doc/mgr/dashboard.rst:626]: Please don't use a continental European accent mark instead of an English apostrophe.

Suggested change to the suggested change:

```
- the user´s browser
+ the user's browser
```

Comment (p-se) [reply doc/mgr/dashboard.rst:626]: :pray:

**Comment by p-se** [comment]
Thank you very much for the suggestions!

File: doc/cephadm/monitoring.rst
Comment (smithfarm) [inline doc/cephadm/monitoring.rst:84]: This is somewhat difficult to parse in English. This would be simpler:

`s/the URL to Grafana which is configured/the Grafana URL configured/`

File: doc/cephadm/monitoring.rst
Comment (smithfarm) [inline doc/cephadm/monitoring.rst:86]: This sentence is awkward, too. How about:

```
One such scenario is when the cluster and the accessing user are each in a different DNS zone.
```

File: doc/mgr/dashboard.rst
Comment (smithfarm) [inline doc/mgr/dashboard.rst:611]: `s/the URL to Grafant/the Grafana URL/`

**Review by smithfarm — DISMISSED** [review]

File: src/pybind/mgr/dashboard/controllers/grafana.py
Comment (epuertat) [inline src/pybind/mgr/dashboard/controllers/grafana.py:30]:
```suggestion
        response = {'instance': str(Settings.GRAFANA_FRONTEND_API_URL or Settings.GRAFANA_API_URL).rstrip('/')}
```
Or, IMHO, even better:
```suggestion
        response = {'instance': (mgr.get_module_option('GRAFANA_FRONTEND_API_URL') or mgr.get_module_option('GRAFANA_API_URL')).rstrip('/')}
```


`self._grafana_frontend_url()` is a thin wrapper around `Settings.GRAFANA_...` which is in turn a thin wrapper around `mgr.get_module_option('GRAFANA_...')`, so by calling a static method in the Grafana controller we're triggering a read to ceph-mgr config option cache... So instead of encapsulating complexity (which there is not) this is just creating a false sense of where information belongs to...

**Review by epuertat — COMMENTED** [review]
I think I already had a loosely related conversation with @sebastian-philipp on dealing with multiple domains for Grafana... I'm personally not a fan of adding more configuration options when other alternatives exist. So, for example, is it possible that Dashboard and end-users use the same short hostname (I'm not even thinking of FQDN) which maps to different IP addresses/subnetworks (via /etc/hosts or nslookup default domain if some DNS server is in place?

**Comment by p-se** [comment]
> I think I already had a loosely related conversation with @sebastian-philipp on dealing with multiple domains for Grafana... I'm personally not a fan of adding more configuration options when other alternatives exist. So, for example, is it possible that Dashboard and end-users use the same short hostname (I'm not even thinking of FQDN) which maps to different IP addresses/subnetworks (via /etc/hosts or nslookup default domain if some DNS server is in place?

I'm frankly not very experienced with that topic, so I cannot fully answer your question. I know that adapting `/etc/hosts` works, but I do not consider that as a user friendly option. I can imagine that a default domain for nslookup might work. But would that not require all clients to be configured properly?

But let's assume that there's an option to make it work by making a configuring at central location, so that it would work for all clients.

Being able to chose between that  and a simple additional option, which can be set once and will then work for all (frontend) users (in a single subnet) does not seem to wrong to me. The implementation of that additional option is trivial and so the effort to maintain it. The gained flexibility is, in my opinion, worth it.

**Comment by epuertat** [comment]
> I'm frankly not very experienced with that topic, so I cannot fully answer your question. I know that adapting `/etc/hosts` works, but I do not consider that as a user friendly option. I can imagine that a default domain for nslookup might work. But would that not require all clients to be configured properly?

@p-se can't Cephadm set this option so we keep this 'internal' rather than force users to manually enter another address when, perhaps, Cephadm would know the public IP address of Grafana instance?

Additionally, and just to keep it in mind, we still have pending the [Grafana HA issue](https://tracker.ceph.com/issues/43668), where one the possibilities was to pass multiple Grafana URLs :/ This change would complicate that... If we could get Cephadm to update this internal option to the active/unique Grafana instance, I think we would minimize the configuration burden.

Last suggestion: given that Grafana dashboards are no longer pushed from Dashboard but built with container images, I'd suggest to stop pinging Grafana from Dashboard... It's almost decoupled except for this bit.

**Comment by p-se** [comment]
> > I'm frankly not very experienced with that topic, so I cannot fully answer your question. I know that adapting `/etc/hosts` works, but I do not consider that as a user friendly option. I can imagine that a default domain for nslookup might work. But would that not require all clients to be configured properly?
> 
> @p-se can't Cephadm set this option so we keep this 'internal' rather than force users to manually enter another address when, perhaps, Cephadm would know the public IP address of Grafana instance?

This setting is by no means forcing the user to do anything. The GRAFANA_API_URL is automatically set by ceph for ceph dashboard. Ceph dashboard will always be able to access Grafana, but if the users of Grafana can use that URL or not, is possibly impossible to determine for ceph. And just in the case where users are, for example, in a different DNS zone, the GRAFANA_FRONTEND_API_URL setting can be used to enable access to Grafana for those users. If you can access Grafana by the URL set by the orchestrator, there's no point in using it, so no force at all. But if you are in a situation where users cannot access Grafana by using the URL determined and set by ceph orchestator, this setting will come in handy.

> 
> Additionally, and just to keep it in mind, we still have pending the [Grafana HA issue](https://tracker.ceph.com/issues/43668), where one the possibilities was to pass multiple Grafana URLs :/ This change would complicate that... If we could get Cephadm to update this internal option to the active/unique Grafana instance, I think we would minimize the configuration burden.

That might always work for the URL the Ceph Dashboard (backend) is supposed to access, but not for the URL the users needs to access Grafana. But you're right, that issue will complicate the matter. Though, what you suggest does not solve the problem of this PR. (But I think it makes sense to think about implementing it for the Grafana HA issue and the GRAFANA_API_URL configuratin option).

> 
> Last suggestion: given that Grafana dashboards are no longer pushed from Dashboard but built with container images, I'd suggest to stop pinging Grafana from Dashboard... It's almost decoupled except for this bit and that would allow to keep a single option (rather than 2 for internal/external routing).

That would surely be a simplification. But even if orchestrator would set a URL for the frontend (well, still in the backend but only used by the frontend), it wouldn't work all the time. The option to overrule this setting by providing an URL for frontend users manually still solves an issue the suggested implementation (possibly) can't.

Thanks for your comments!

**Comment by sebastian-philipp** [comment]
cephadm just does not control the subject of TLS certificates. Can be anything: IPs, different domain names, different host names. We just can't demand that the ceph's internal host names used by the cluster match the TLS certificate of Grafana.

**Comment by sebastian-philipp** [comment]
I also see a potential UX issue: https://tracker.ceph.com/issues/46304 We need to inform users that they should use a different config option then.

**Review by epuertat — CHANGES_REQUESTED** [review]
@p-se After talking with some RH field folks, it seems that the case of having Dashboard, Grafana, etc.,  behind a reverse proxy is rather common, so being that the case, why not having a single `proxy_origin` or `proxy_base_url` (since it'd also include protocol + hostname + port... [[reference]](https://stackoverflow.com/questions/37232382/what-is-protocol-and-host-combined-called)), and not only for Grafana, but also for Dashboard itself (e.g.: for the Active-Standby redirection or for exposing REST API docs)?

In any case I think this should fix the `validation` endpoint too, as it's called from the front-end, so it means that this will be broken as well.

**Comment by p-se** [comment]
> @p-se After talking with some RH field folks, it seems that the case of having Dashboard, Grafana, etc., behind a reverse proxy is rather common, so being that the case, why not having a single `proxy_origin` or `proxy_base_url` (since it'd also include protocol + hostname + port... [[reference]](https://stackoverflow.com/questions/37232382/what-is-protocol-and-host-combined-called)), and not only for Grafana, but also for Dashboard itself (e.g.: for the Active-Standby redirection, for exposing R[EST API docs](https://tracker.ceph.com/issues/48252), or for [SSO](https://tracker.ceph.com/issues/48306))?

Do I understand correctly and you suggest to rename the additional configuration option to something more generic, to be able to use it for other components as well (especially when a reverse proxy is configured)? 

> 
> In any case I think this should fix the `validation` endpoint too, as it's called from the front-end, so it means that this will be broken as well.

It is indeed called from the frontend, but the backend checks if the Grafana graphs exist, so the backend needs to have the original and automatically determinable host configured. I don't see a relation to the new configuration option.

**Review by votdev — APPROVED** [review]

**Review by epuertat — APPROVED** [review]
In order to speed up this, I'll give my ok here, but I'd ask you to provide a solution for the single issue I think we shouldn't merge this with: a user specifying `GRAFANA_API_URL` and `GRAFANA_FRONTEND_API_URL` to different protocols and ports. Ideas:
* Ensure that HTTP protocol and ports match when the user enters the `set-grafana-...` command, and fail otherwise.
* Specify only a FQDN (instead of a URL) in a `GRAFANA_FRONTEND_API_FQDN` and reuse the proto and port from `GRAFANA_API_URL`.

I opened a separate ticket to get rid of the [backend-Grafana validation behaviour](https://tracker.ceph.com/issues/48384).

File: src/pybind/mgr/dashboard/controllers/grafana.py
Comment (p-se) [reply src/pybind/mgr/dashboard/controllers/grafana.py:25]: I'm not sure if we want to tell users that they should use TLS certificates for IPs, which we had to if we went with the "Use IP addresses directly" proposal. PQDNs would also need to be set correctly by cephadm. If cephadm is not capable of setting it correctly, the user will need an alternative option to set it manually.

> cephadm just does not control the subject of TLS certificates. Can be anything: IPs, different domain names, different host names. We just can't demand that the ceph's internal host names used by the cluster match the TLS certificate of Grafana.
>
> Not saying anything about this implementation here, but there has to be some way for users to make the dashboard embed Grafana as defined in the certificate.

Comment (epuertat) [reply src/pybind/mgr/dashboard/controllers/grafana.py:25]: Mmm so, in this case, whoever configuring the certs (cephadm or a manual operator) should know exactly what's the 'user-facing' URL?

I don't know exactly what's the best approach here, since I feel the problem is a bit undefined, but if we all agree on that the existing Dashboard backend-Grafana interaction is unnecessary, then we can free that configuration option and use it for this purpose (or simply provide a check to ensure the front-end and back-end options are not incompatible)...

Comment (p-se) [reply src/pybind/mgr/dashboard/controllers/grafana.py:25]: I don't think cephadm will ever be able to always determine the correct user-facing URL, neither as it is now nor with the knowledge  of the certificate.

In my humble opinion the best approach is to simply continue merging this PR (when I fixed the tests of course) and then to discuss different open topics around monitoring that do affect how the feature in this PR might needed to be changed. Like the re-evaluation of the reverse proxy for Grafana or the high availability. Each of these topics can change this implementation and by only looking at this implementation, I won't be able to determine the best implementation.

**Comment by p-se** [comment]
jenkins test make check

**Comment by p-se** [comment]
jenkins test make check

**Comment by p-se** [comment]
jenkins test make check

**Comment by votdev** [comment]
jenkins test make check

**Comment by votdev** [comment]
jenkins test dashboard

File: src/pybind/mgr/dashboard/controllers/grafana.py
Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard/controllers/grafana.py:25]: I'm feeling like a broken record now. cephadm can't always determine the correct thing that the user needs to point to, in order for user's browser to reach Grafana. It just can't.

* Might be that the Grafana certificate has a specific common name
* Might be that the user need to access Grafana via a proxy
* Might be that a user need to use an FQDN to access Grafana from outside of the cluster, while the cluster itself is configured without DNS. 

Therefor users need to be able to set the grafana URL to some custom name. 

I mean, we already have a document that covers exactly that:

https://docs.ceph.com/en/latest/mgr/dashboard/#configuring-dashboard

I still don't understand why we must break user's expectations here.

Comment (epuertat) [reply src/pybind/mgr/dashboard/controllers/grafana.py:25]: @sebastian-philipp: so we are expecting that, after the cluster is deployed with a deployment tool (cephadm), users will still need to perform several manual steps involving setting IP addresses, hostnames, etc. just for the dashboard to simply work?

I don't think that's user-friendly at all, and at least it doesn't fill the feature gap from previous Ceph deployment tools (I can only speak for Ceph-Ansible, which for example had a [`dashboard_frontend_vip` setting](https://github.com/ceph/ceph-ansible/blob/master/roles/ceph-dashboard/tasks/configure_dashboard.yml#L185) for configuring a front-end FQDN). I don't think it's correct to embed complexity within the dashboard, when the complexity can be dealt with from the 'configuration' tool.

That said, and given this is affecting a downstream product, I already gave my ok to this PR, and just asked @p-se to add a fix so these 2 options don't allow users to make mistakes by entering incorrect data (e.g.: by configuring different HTTP schemes or ports for the same Grafana endpoint).

Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard/controllers/grafana.py:25]: > @sebastian-philipp: so we are expecting that, after the cluster is deployed with a deployment tool (cephadm), users will still need to perform several manual steps involving setting IP addresses, hostnames, etc. just for the dashboard to simply work?

The problem is, cephadm can't always determine the correct thing. It's simply outside of the control of cephadm. cephadm doesn't mandate a working DNS. cephadm doesn't mandate which url a user need to enter to access Grafana. cephadm doesn't mandate that a hostname must be a FQDN. cephadm doesn't mandate that host names must be bare. it's simply that cephadm doesn't care, as long as we can properly connect to the host. How uses access Gafana is something that cephadm doesn't care about. Sure, we can provide a good default, but that is not necessarily working for all users.

>  Ceph-Ansible, which for example had a dashboard_frontend_vip setting for configuring a front-end FQDN).

Exactly! ceph-ansible exposes a *user defined* setting `dashboard_frontend_vip` to the user to allow them to change it and is simply forwarding this to `ceph dashboard set-grafana-api-url`, because ceph-ansible has the same problem as cephadm. In constrast to ceph-ansible, cephadm doesn't provide a `dashboard_frontend_vip` setting yet, and instead let users change the grafana url themselves.

Comment (epuertat) [reply src/pybind/mgr/dashboard/controllers/grafana.py:25]: > The problem is, cephadm can't always determine the correct thing.

Mmm, I never said that cephadm _had to guess_ this value from scratch, just that I think it should set it. Where from? Perhaps it could be inferred from another deployment setting or be directly set from a `dashboard_frontend_vip`-like option... 

My main concern about this PR was that it's providing an ad-hoc fix while leaving some other issues behind. So we'd later need to create `MGR_REDIRECT_FRONTEND_URL`, `SSO_FRONTEND_URL` and `OPENAPI_DOCS_URL` with probably the exact same data.

However, the dashboard simply guesses its hostname from the `socket.getfqdn()`, which in turn looks up `/etc/hosts`, so perhaps the [approach described here](https://tracker.ceph.com/issues/46304#note-7) might perfectly work for all those cases without the need of these changes.

Comment (p-se) [reply src/pybind/mgr/dashboard/controllers/grafana.py:25]: > That said, and given this is affecting a downstream product, I already gave my ok to this PR, and just asked @p-se to add a fix so these 2 options don't allow users to make mistakes by entering incorrect data (e.g.: by configuring different HTTP schemes or ports for the same Grafana endpoint).

Isn't it so that if a reverse proxy is configured neither the scheme nor the port needs to be identical with the URL the dashboard backend needs to use compared with the one frontend users would need to use?

If you agree, that for the reverse proxy to work with this setting, we cannot implement the check you've requested, can we get this PR merged then?

Comment (epuertat) [reply src/pybind/mgr/dashboard/controllers/grafana.py:25]: Yes, you're right on that. I still miss fixing the whole issue, but as agreed, let's go and merge this one when the dashboard tests are passing.

**Comment by epuertat** [comment]
jenkins test dashboard

**Comment by epuertat** [comment]
dashboard tests:
```
Get:1 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]
Hit:2 http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu bionic InRelease
Get:3 https://dl.google.com/linux/chrome/deb stable InRelease [1811 B]
Ign:4 http://deb.rug.nl/ppa/mirror/ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu bionic InRelease
Err:5 http://deb.rug.nl/ppa/mirror/ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu bionic Release
  404  Not Found [IP: 129.125.36.99 80]
Hit:6 http://nova.clouds.archive.ubuntu.com/ubuntu bionic InRelease
Get:7 http://nova.clouds.archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]
Hit:8 http://mirror.nullivex.com/ppa/ubuntu-toolchain-r-test bionic InRelease
Ign:9 https://chacra.ceph.com/r/libboost/master/7aba8a1882670522ee1d1ee1bba0ea170b292dec/ubuntu/bionic/flavors/default bionic InRelease
Get:10 https://dl.google.com/linux/chrome/deb stable/main amd64 Packages [1091 B]
Hit:11 https://chacra.ceph.com/r/libboost/master/7aba8a1882670522ee1d1ee1bba0ea170b292dec/ubuntu/bionic/flavors/default bionic Release
Get:12 http://nova.clouds.archive.ubuntu.com/ubuntu bionic-backports InRelease [74.6 kB]
Ign:13 https://chacra.ceph.com/r/libboost/master/7aba8a1882670522ee1d1ee1bba0ea170b292dec/ubuntu/bionic/flavors/default bionic Release.gpg
Reading package lists...
E: The repository 'http://deb.rug.nl/ppa/mirror/ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu bionic Release' does not have a Release file.
```

jenkins test dashboard

**Comment by epuertat** [comment]
jenkins test dashboard

**Comment by callithea** [comment]
- [pkg build](https://shaman.ceph.com/builds/ceph/wip-laura-testing-37937-38534/)
- [QA run](https://pulpito.ceph.com/laura-2020-12-15_12:38:21-rados:dashboard-wip-laura-testing-37937-38534-distro-basic-smithi/) [WIP]

**Review by callithea — APPROVED** [review]
Lgtm

**Pull request merged**
